inspirehep/modules/disambiguation/core/ml/models.py

```diff
@@ -76,3 +101,174 @@ def fit(self):
+                    ('keywords', FuncTransformer(func=get_keywords)),
+                    ('shaper', Shaper(newshape=(-1,))),
+                    ('tf-idf', TfidfVectorizer(dtype=np.float32,
+                                               decode_error='replace')),
+                ]), groupby=group_by_signature)),
+                ('combiner', CosineSimilarity())
+            ])),
+            ('collaborations_similarity', Pipeline([
+                ('pairs', PairTransformer(element_transformer=Pipeline([
+                    ('collaborations', FuncTransformer(
+                        func=get_collaborations)),
+                    ('shaper', Shaper(newshape=(-1,))),
+                    ('tf-idf', TfidfVectorizer(dtype=np.float32,
+                                               decode_error='replace')),
+                ]), groupby=group_by_signature)),
+                ('combiner', CosineSimilarity())
+            ])),
+            ('subject_similairty', Pipeline([
+                ('pairs', PairTransformer(element_transformer=Pipeline([
+                    ('keywords', FuncTransformer(func=get_topics)),
+                    ('shaper', Shaper(newshape=(-1))),
+                    ('tf-idf', TfidfVectorizer(dtype=np.float32,
+                                               decode_error='replace')),
+                ]), groupby=group_by_signature)),
+                ('combiner', CosineSimilarity())
+            ])),
+            ('title_similarity', Pipeline([
+                ('pairs', PairTransformer(element_transformer=Pipeline([
+                    ('title', FuncTransformer(func=get_title)),
+                    ('shaper', Shaper(newshape=(-1,))),
+                    ('tf-idf', TfidfVectorizer(analyzer='char_wb',
+                                               ngram_range=(2, 4),
+                                               dtype=np.float32,
+                                               decode_error='replace')),
+                ]), groupby=group_by_signature)),
+                ('combiner', CosineSimilarity())
+            ])),
+            ('author_ethnicity', Pipeline([
+                ('pairs', PairTransformer(element_transformer=Pipeline([
+                    ('name', FuncTransformer(func=get_author_full_name)),
+                    ('shaper', Shaper(newshape=(-1,))),
+                    ('classifier', EstimatorTransformer(self.ethnicity_estimator.estimator)),
+                ]), groupby=group_by_signature)),
+                ('sigmoid', FuncTransformer(func=expit)),
+                ('shaper', Shaper(newshape=(-1, 2))),
+                ('combiner', ElementMultiplication())
+            ]))
+        ])
+        classifier = RandomForestClassifier(n_estimators=500, n_jobs=8)
+
+        self.distance_estimator = Pipeline([('transformer', transformer), ('classifier', classifier)])
+        self.distance_estimator.fit(self.X, self.y)
+
+    def predict(self, X):
+        return self.distance_estimator.predict(X)
```

Comment:
General comment: I would be more explicit about the state of the model here.
On general grounds, you want to either `fit` or `predict`. In the first case, you need to `load_data` before and `save_model` after. In the second case, you need to `load_model` and call `predict`. However, none of this is enforced by the class, you can happily do nonsense combinations like `predict` without `load_model`. It might make sense to enforce this doesn't happen.

Reply:
This is the idea I discarded in https://github.com/inspirehep/inspire-next/pull/3443... I preferred going with an API method such as https://github.com/inspirehep/inspire-next/blob/87a00d2503efb490b16c992671eeb92ac4410938/inspirehep/modules/disambiguation/api.py#L103-L108 that enforces a certain order rather than enforce it at the decorator level. If you think it was a good idea after all I can resurrect it!